roachtest: tpchvec/disk failed

[(roachtest).tpchvec/disk failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1814553&tab=buildLog) on [master@b5f030223fbcf22e806c48a3c46e74a73a54f50f](https://github.com/cockroachdb/cockroach/commits/b5f030223fbcf22e806c48a3c46e74a73a54f50f):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20200318-1814553/tpchvec/disk/run_1
	tpchvec.go:579,tpchvec.go:669,test_runner.go:753: error with attached stack trace:
		    main.execCmdWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:426
		    main.(*cluster).RunWithBuffer
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster.go:2031
		    main.registerTPCHVec.func1
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:574
		    main.registerTPCHVec.func3
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/tpchvec.go:669
		    main.(*testRunner).runTest.func2
		    	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test_runner.go:753
		    runtime.goexit
		    	/usr/local/go/src/runtime/asm_amd64.s:1357
		  - error with embedded safe details: %s
		    -- arg 1: <string>
		  - /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1814553-1584515366-86-n3cpu4:1 -- ./workload run tpch --concurrency=1 --db=tpch --max-ops=1 --queries=3 --vectorize=on {pgurl:1-3}:
		  - exit status 1

	cluster.go:1410,context.go:135,cluster.go:1399,test_runner.go:825: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-1814553-1584515366-86-n3cpu4 --oneshot --ignore-empty-nodes: exit status 1 2: 5968
		1: dead
		3: 5848
		Error:  1: dead

```

<details><summary>More</summary><p>


Artifacts: [/tpchvec/disk](https://teamcity.cockroachdb.com/viewLog.html?buildId=1814553&tab=artifacts#/tpchvec/disk)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Atpchvec%2Fdisk.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0xa0 pc=0x2294467]

goroutine 8418 [running]:
panic(0x3db1700, 0x6ddbe10)
	/usr/local/go/src/runtime/panic.go:722 +0x2c2 fp=0xc0042a7b88 sp=0xc0042a7af8 pc=0x79a432
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:199
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:394 +0x3ec fp=0xc0042a7bb8 sp=0xc0042a7b88 pc=0x7b004c
github.com/cockroachdb/cockroach/pkg/sql/colcontainer.(*diskQueue).Close(0x0, 0x4b7f880, 0xc005942340, 0xc0042a7e08, 0xc0004d6380)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colcontainer/diskqueue.go:398 +0x37 fp=0xc0042a7c90 sp=0xc0042a7bb8 pc=0x2294467
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*spillingQueue).close(0xc0046083c0, 0x4b7f880, 0xc005942340, 0xc0042a7d28, 0xc0042a7d50)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/spilling_queue.go:263 +0x73 fp=0xc0042a7cc8 sp=0xc0042a7c90 pc=0x2944ae3
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).closeLocked(0xc0044bb440, 0x4b7f880, 0xc005942340)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:225 +0x4b fp=0xc0042a7d30 sp=0xc0042a7cc8 pc=0x2853b6b
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*routerOutputOp).cancel(0xc0044bb440, 0x4b7f880, 0xc005942340)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:238 +0x5f fp=0xc0042a7d60 sp=0xc0042a7d30 pc=0x2853c5f
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run.func1(0x4afaf20, 0xc004d8ab50)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:456 +0x18c fp=0xc0042a7e50 sp=0xc0042a7d60 pc=0x29dc77c
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run(0xc0044bb7a0, 0x4b7f880, 0xc005942340)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:496 +0x1b6 fp=0xc0042a7f38 sp=0xc0042a7e50 pc=0x2854b96
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRouter.func1(0x4b7f880, 0xc005942340, 0xc004de4170)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:605 +0xa5 fp=0xc0042a7f90 sp=0xc0042a7f38 pc=0x2b4acb5
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc004e2b480, 0x4b7f880, 0xc005942340, 0xc004de4170, 0xc0042341f0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1090 +0x44 fp=0xc0042a7fb8 sp=0xc0042a7f90 pc=0x2b4af74
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc0042a7fc0 sp=0xc0042a7fb8 pc=0x7caaa1
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1089 +0x6f
```

--- a/pkg/sql/colexec/routers_test.go
+++ b/pkg/sql/colexec/routers_test.go
@@ -402,9 +402,15 @@ func TestRouterOutputRandom(t *testing.T) {
 
 				expected := make(tuples, 0, len(data))
 
+				// canceled is a boolean that specifies whether the output was canceled.
+				// If this is the case, the output should not be verified.
+				canceled := false
+
 				// Producer.
 				errCh := make(chan error)
+				wg.Add(1)
 				go func() {
+					defer wg.Done()
 					lastBlockedState := false
 					for {
 						b := inputs[0].Next(ctx)
@@ -432,6 +438,13 @@ func TestRouterOutputRandom(t *testing.T) {
 								}
 							}
 							lastBlockedState = true
+						}
+
+						if rng.Float64() < 0.1 {
+							o.cancel(ctx)
+							canceled = true
+							errCh <- nil
+							return
 						}
 
 						// Read any state changes.
@@ -480,6 +493,10 @@ func TestRouterOutputRandom(t *testing.T) {
 				}
 
 				wg.Wait()
+
+				if canceled {
+					return
+				}
 
 				if err := newOpTestOutput(actual, expected).Verify(); err != nil {
 					t.Fatal(err)

--- a/pkg/sql/colexec/spilling_queue.go
+++ b/pkg/sql/colexec/spilling_queue.go
@@ -48,6 +48,7 @@ type spillingQueue struct {
 	curTailIdx       int
 	numInMemoryItems int
 	numOnDiskItems   int
+	closed           bool
 
 	diskQueueCfg   colcontainer.DiskQueueCfg
 	diskQueue      colcontainer.Queue
@@ -235,12 +236,19 @@ func (q *spillingQueue) maybeSpillToDisk(ctx context.Context) error {
 		}
 	}
 	log.VEvent(ctx, 1, "spilled to disk")
+	var diskQueue colcontainer.Queue
 	if q.rewindable {
-		q.diskQueue, err = colcontainer.NewRewindableDiskQueue(ctx, q.typs, q.diskQueueCfg, q.diskAcc)
+		diskQueue, err = colcontainer.NewRewindableDiskQueue(ctx, q.typs, q.diskQueueCfg, q.diskAcc)
 	} else {
-		q.diskQueue, err = colcontainer.NewDiskQueue(ctx, q.typs, q.diskQueueCfg, q.diskAcc)
+		diskQueue, err = colcontainer.NewDiskQueue(ctx, q.typs, q.diskQueueCfg, q.diskAcc)
 	}
-	return err
+	if err != nil {
+		return err
+	}
+	// Only assign q.diskQueue if there was no error, otherwise the returned value
+	// may be non-nil but invalid.
+	q.diskQueue = diskQueue
+	return nil
 }
 
 // empty returns whether there are currently no items to be dequeued.
@@ -256,11 +264,18 @@ func (q *spillingQueue) spilled() bool {
 }
 
 func (q *spillingQueue) close(ctx context.Context) error {
+	if q.closed {
+		return nil
+	}
 	if q.diskQueue != nil {
+		if err := q.diskQueue.Close(ctx); err != nil {
+			return err
+		}
 		if q.fdSemaphore != nil {
 			q.fdSemaphore.Release(q.numFDsOpenAtAnyGivenTime())
 		}
-		return q.diskQueue.Close(ctx)
+		q.closed = true
+		return nil
 	}
 	return nil
 }
@@ -284,6 +299,7 @@ func (q *spillingQueue) reset(ctx context.Context) {
 		execerror.VectorizedInternalPanic(err)
 	}
 	q.diskQueue = nil
+	q.closed = false
 	q.numInMemoryItems = 0
 	q.numOnDiskItems = 0
 	q.curHeadIdx = 0

--- a/pkg/sql/colflow/vectorized_flow.go
+++ b/pkg/sql/colflow/vectorized_flow.go
@@ -37,6 +37,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
@@ -104,10 +105,13 @@ type vectorizedFlow struct {
 
 	tempStorage struct {
 		// path is the path to this flow's temporary storage directory.
-		path string
-		// created is an atomic that is set to 1 when the flow's temporary storage
-		// directory has been created.
-		created int32
+		path           string
+		createdStateMu struct {
+			syncutil.Mutex
+			// created is a protected boolean that is true when the flow's temporary
+			// storage directory has been created.
+			created bool
+		}
 	}
 
 	testingKnobs struct {
@@ -162,14 +166,17 @@ func (f *vectorizedFlow) Setup(
 		FS:   f.Cfg.TempFS,
 		Path: f.tempStorage.path,
 		OnNewDiskQueueCb: func() {
-			if !atomic.CompareAndSwapInt32(&f.tempStorage.created, 0, 1) {
+			f.tempStorage.createdStateMu.Lock()
+			defer f.tempStorage.createdStateMu.Unlock()
+			if f.tempStorage.createdStateMu.created {
 				// The temporary storage directory has already been created.
 				return
 			}
-			log.VEventf(ctx, 2, "flow %s spilled to disk, stack trace: %s", f.ID, util.GetSmallTrace(2))
+			log.VEventf(ctx, 1, "flow %s spilled to disk, stack trace: %s", f.ID, util.GetSmallTrace(2))
 			if err := f.Cfg.TempFS.CreateDir(f.tempStorage.path); err != nil {
 				execerror.VectorizedInternalPanic(errors.Errorf("unable to create temporary storage directory: %v", err))
 			}
+			f.tempStorage.createdStateMu.created = true
 		},
 	}
 	if err := diskQueueCfg.EnsureDefaults(); err != nil {
@@ -279,7 +286,10 @@ func (f *vectorizedFlow) Cleanup(ctx context.Context) {
 		mon.Stop(ctx)
 	}
 
-	if atomic.LoadInt32(&f.tempStorage.created) == 1 {
+	f.tempStorage.createdStateMu.Lock()
+	created := f.tempStorage.createdStateMu.created
+	f.tempStorage.createdStateMu.Unlock()
+	if created {
 		if err := f.tryRemoveAll(f.tempStorage.path); err != nil {
 			// Log error as a Warning but keep on going to close the memory
 			// infrastructure.

--- a/pkg/sql/colflow/vectorized_flow_test.go
+++ b/pkg/sql/colflow/vectorized_flow_test.go
@@ -12,9 +12,11 @@ package colflow
 
 import (
 	"context"
+	"path/filepath"
 	"sync"
 	"testing"
 
+	"github.com/cockroachdb/cockroach/pkg/base"
 	"github.com/cockroachdb/cockroach/pkg/col/coltypes"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
@@ -27,6 +29,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/stretchr/testify/require"
 )
@@ -235,11 +238,13 @@ func TestVectorizedFlowTempDirectory(t *testing.T) {
 	ctx := context.Background()
 	defer evalCtx.Stop(ctx)
 
-	const baseDirName = "base"
-
-	ngn := storage.NewDefaultInMem()
+	// We use an on-disk engine for this test since we're testing FS interactions
+	// and want to get the same behavior as a non-testing environment.
+	tempPath, dirCleanup := testutils.TempDir(t)
+	ngn, err := storage.NewDefaultEngine(0 /* cacheSize */, base.StorageConfig{Dir: tempPath})
+	require.NoError(t, err)
 	defer ngn.Close()
-	require.NoError(t, ngn.CreateDir(baseDirName))
+	defer dirCleanup()
 
 	newVectorizedFlow := func() *vectorizedFlow {
 		return NewVectorizedFlow(
@@ -247,7 +252,7 @@ func TestVectorizedFlowTempDirectory(t *testing.T) {
 				FlowCtx: execinfra.FlowCtx{
 					Cfg: &execinfra.ServerConfig{
 						TempFS:          ngn,
-						TempStoragePath: baseDirName,
+						TempStoragePath: tempPath,
 						VecFDSemaphore:  &colexec.TestingSemaphore{},
 						Metrics:         &execinfra.DistSQLMetrics{},
 					},
@@ -258,11 +263,15 @@ func TestVectorizedFlowTempDirectory(t *testing.T) {
 		).(*vectorizedFlow)
 	}
 
-	checkDirs := func(t *testing.T, numDirs int) {
+	dirs, err := ngn.ListDir(tempPath)
+	require.NoError(t, err)
+	numDirsTheTestStartedWith := len(dirs)
+	checkDirs := func(t *testing.T, numExtraDirs int) {
 		t.Helper()
-		dirs, err := ngn.ListDir(baseDirName)
+		dirs, err := ngn.ListDir(tempPath)
 		require.NoError(t, err)
-		require.Equal(t, numDirs, len(dirs), "expected %d directories but found %d: %s", numDirs, len(dirs), dirs)
+		expectedNumDirs := numDirsTheTestStartedWith + numExtraDirs
+		require.Equal(t, expectedNumDirs, len(dirs), "expected %d directories but found %d: %s", expectedNumDirs, len(dirs), dirs)
 	}
 
 	// LazilyCreated asserts that a directory is not created during flow Setup
@@ -339,6 +348,31 @@ func TestVectorizedFlowTempDirectory(t *testing.T) {
 		vf1.Cleanup(ctx)
 		checkDirs(t, 1)
 		vf2.Cleanup(ctx)
+		checkDirs(t, 0)
+	})
+
+	t.Run("DirCreationRace", func(t *testing.T) {
+		vf := newVectorizedFlow()
+		var creator *vectorizedFlowCreator
+		vf.testingKnobs.onSetupFlow = func(c *vectorizedFlowCreator) {
+			creator = c
+		}
+
+		_, err := vf.Setup(ctx, &execinfrapb.FlowSpec{}, flowinfra.FuseNormally)
+		require.NoError(t, err)
+
+		createTempDir := creator.diskQueueCfg.OnNewDiskQueueCb
+		errCh := make(chan error)
+		go func() {
+			createTempDir()
+			errCh <- ngn.CreateDir(filepath.Join(vf.tempStorage.path, "async"))
+		}()
+		createTempDir()
+		// Both goroutines should be able to create their subdirectories within the
+		// flow's temporary directory.
+		require.NoError(t, ngn.CreateDir(filepath.Join(vf.tempStorage.path, "main_goroutine")))
+		require.NoError(t, <-errCh)
+		vf.Cleanup(ctx)
 		checkDirs(t, 0)
 	})
 }

